src/NServiceBus.Core.Analyzer/AwaitOrCaptureTasksAnalyzer.cs

```diff
@@ -0,0 +1,90 @@
+namespace NServiceBus.Core.Analyzer
+{
+    using System;
+    using System.Collections.Immutable;
+    using System.Linq;
+    using Microsoft.CodeAnalysis;
+    using Microsoft.CodeAnalysis.CSharp;
+    using Microsoft.CodeAnalysis.CSharp.Syntax;
+    using Microsoft.CodeAnalysis.Diagnostics;
+
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class AwaitOrCaptureTasksAnalyzer : DiagnosticAnalyzer
+    {
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(diagnostic);
+
+        public override void Initialize(AnalysisContext context) => context.RegisterSyntaxNodeAction(Analyze, SyntaxKind.InvocationExpression);
+
+        void Analyze(SyntaxNodeAnalysisContext context)
+        {
+            if (!(context.Node is InvocationExpressionSyntax call))
+            {
+                return;
+            }
+
+            if (!(call.Parent is ExpressionStatementSyntax))
+            {
+                return;
+            }
+
+            foreach (var token in call.Expression?.DescendantTokens() ?? Enumerable.Empty<SyntaxToken>())
+            {
+                // check syntax tree (cheap) first for possible call requiring await and then check semantic model (expensive) to confirm
+                if (CouldBeMethodRequiringAwait(token) && IsMethodRequiringAwait(call, context))
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(diagnostic, call.GetLocation(), call.ToString()));
+                    return;
+                }
+            }
+        }
+
+        static bool CouldBeMethodRequiringAwait(SyntaxToken syntaxToken) =>
+            syntaxToken.Kind() == SyntaxKind.IdentifierToken && methodNames.Contains(syntaxToken.Text);
+
+        static bool IsMethodRequiringAwait(ExpressionSyntax call, SyntaxNodeAnalysisContext context) =>
+            context.SemanticModel.GetSymbolInfo(call).Symbol is IMethodSymbol methodSymbol &&
+                methods.Contains(methodSymbol.GetFullName());
+
+        static readonly DiagnosticDescriptor diagnostic = new DiagnosticDescriptor(
+            "NSB0001",
+            "Await or capture tasks",
```

Comment:
Is "capture" the correct word here? perhaps "assign" would be better? Also, should "tasks" be capitalized?

Reply:
> Is "capture" the correct word here? perhaps "assign" would be better?

I'm not entirely sure. The thing is, if we do `foo.Bar().ConfigureAwait(false)` that also prevents the diagnostic from firing for `Bar()`, but what is the best description of that? I wouldn't say it's an "assignment".

> Also, should "tasks" be capitalized?

I'm not sure what the convention is for pluralising identifiers. I.e. the identifier is `Task`, and I'd expect that to be monospaced in rich text (even though rich text is not supported in this context). Would it be correct to pluralise the identifier to `Tasks`? Or is "task" a _conceptual_ reference, which happens to implemented in .NET by the `Task` type?

Or maybe we can reword the title to avoid the plural?

Reply:
> I'm not entirely sure. The thing is, if we do foo.Bar().ConfigureAwait(false) that also prevents the diagnostic from firing for Bar(), but what is the best description of that? I wouldn't say it's an "assignment".

If that passes the analyzer, then I'd argue that the analyzer might be too simple and that it should detect that case as well.

Reply:
`Await or capture Task` sounds reasonable to me?

> If that passes the analyzer, then I'd argue that the analyzer might be too simple and that it should detect that case as well.

that was discussed and we decided to not try to capture this scenario. That's something to bring up on the pdev issue if you think we need to capture this case.